datafusion-federation/src/analyzer/mod.rs

```diff
@@ -403,35 +427,20 @@ impl FederationProvider for NopFederationProvider {
 }
 
 /// Recursively find the [`FederationProvider`] for all [`TableReference`] instances in the plan.
-/// This information is used to resolve the federation provider for [`Expr::OuterReferenceColumn`].
+/// This is used to resolve the federation providers for [`Expr::OuterReferenceColumn`].
 fn get_plan_provider_recursively(
     plan: &LogicalPlan,
-) -> Result<HashMap<TableReference, ScanResult>> {
-    let mut providers: HashMap<TableReference, ScanResult> = HashMap::new();
-
-    plan.apply(&mut |p: &LogicalPlan| -> Result<TreeNodeRecursion> {
-        // LogicalPlan::SubqueryAlias can also be referred by OuterReferenceColumn
-        // Get the federation provider for TableReference representing LogicalPlan::SubqueryAlias
-        if let LogicalPlan::SubqueryAlias(a) = p {
-            let subquery_alias_providers = get_plan_provider_recursively(&Arc::clone(&a.input))?;
-            let mut provider: ScanResult = ScanResult::None;
-            for (_, i) in subquery_alias_providers {
-                provider.merge(i);
-            }
-            providers.insert(a.alias.clone(), provider);
-        }
+) -> Result<HashMap<TableReference, Arc<dyn FederationProvider>>> {
+    let mut providers: HashMap<TableReference, Arc<dyn FederationProvider>> = HashMap::new();
 
+    plan.apply_with_subqueries(&mut |p: &LogicalPlan| -> Result<TreeNodeRecursion> {
```

Comment:
By using `apply_with_subqueries` we don't need to manually recurse into subqueries, DataFusion handles it for us.  ref https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.LogicalPlan.html#method.apply_with_subqueries